resource-managers/kubernetes/architecture-docs/submission-client.md

```diff
@@ -0,0 +1,220 @@
+---
+layout: global
+title: Implementation of Submitting Applications to Kubernetes
+---
+
+
+Similarly to YARN and Standalone mode, it is common for Spark applications to be deployed on Kubernetes through the
+`spark-submit` process. Applications are deployed on Kubernetes via sending YML files to the Kubernetes API server.
+These YML files declare the structure and behavior of the processes that will be run. However, such a declarative
+approach to application deployment differs considerably from how Spark applications are deployed via the `spark-submit`
+API. There are contracts provided by `spark-submit` that should work in Kubernetes in a consistent manner to the other
+cluster managers that `spark-submit` can deploy on.
+
+This document outlines the design of the **Kubernetes submission client**, which effectively serves as a *translation
+of options provided in spark-submit to a specification of one or more Kubernetes API resources that represent the
+Spark driver*.
+
+# Entry Point
+
+As with the other cluster managers, the user's invocation of `spark-submit` will eventually delegate to running
+`org.apache.spark.deploy.SparkSubmit#submit`. This method calls a main method that handles the submission logic
+for a specific type of cluster manager. The top level entry point for the Kubernetes submission logic is in
+`org.apache.spark.deploy.kubernetes.submit.Client#main()`.
+
+# Driver Configuration Steps
+
+Submitting applications to Kubernetes became more and more complex over time as the submission client had to accommodate
+the large feature matrices of the Spark and Kubernetes ecosystems. The submission client is thus broken down into a
+series of configuration steps, each of which is responsible for handling some specific aspect of configuring the driver.
+A top level component then iterates through all of the steps to produce a final set of Kubernetes resources that are
+then deployed on the cluster.
+
+## Interface Definitions
+
+More formally, a configuration step must implement the following trait:
+
+    package org.apache.spark.deploy.kubernetes.submit.submitsteps
+
+    /**
+     * Represents a step in preparing the Kubernetes driver.
+     */
+    private[spark] trait DriverConfigurationStep {
```

Comment:
My 2cents around putting interface code in arch docs, is that it's probably easier for the maintainers if we document to this detail once we move out of beta. Otherwise I will suspect it's a constant moving target where we're not really maintaining this interface as general public API